### PR TITLE
add vhost-net support

### PR DESCRIPTION
**Pull request by fkcrazy001**
1. add vhost-net as kni support
2. bugfix of kni queue racing
3. del KNI_ENABLE macro, because we can support kni in all cases

### CODE REVIEW

**Comment by pengjianzhang** [comment]
1. #include <rte_kni.h>   会不会在高版本DPDK编译不过？
2. 注释保持C风格
3. 局部变量保持在函数前面定义

逻辑还没细看，待补充

**Comment by fkcrazy001** [comment]
1. #include <rte_kni.h> 已修改为只在 kni.c 中包含， vhost-net 模块不需要，之前是从kni.c中拷贝过来忘记删除了
2-3 已修改

**Comment by fkcrazy001** [comment]
> 1. #include <rte_kni.h>   会不会在高版本DPDK编译不过？
> 2. 注释保持C风格
> 3. 局部变量保持在函数前面定义
> 
> 逻辑还没细看，待补充

1. #include <rte_kni.h> 已修改为只在 kni.c 中包含， vhost-net 模块不需要，之前是从kni.c中拷贝过来忘记删除了
2-3 已修改

File: src/port.h [hunk 1]
```diff
@@ -67,9 +66,8 @@ struct netif_port {
     struct ip_range client_ip_range; /* only used by client; server use all client ip range */
     struct ip_range server_ip_range;
 
-#ifdef KNI_ENABLE
-    struct rte_kni *kni;
-#endif
+    void *kni;
+    struct rte_ring *kni_ring;
```
Comment (pengjianzhang) [inline src/port.h:70]: kni_ring -> kni_mbuf_queue

File: src/port.h [hunk 2]
```diff
@@ -67,9 +66,8 @@ struct netif_port {
     struct ip_range client_ip_range; /* only used by client; server use all client ip range */
     struct ip_range server_ip_range;
 
-#ifdef KNI_ENABLE
-    struct rte_kni *kni;
-#endif
+    void *kni;
```
Comment (pengjianzhang) [inline src/port.h:69]: /* kni or virtio_user */

File: src/virtio-user.c [hunk 3]
```diff
@@ -0,0 +1,372 @@
...
+#define RTE_ETHER_ADDR_BYTES(mac_addrs)	 ((mac_addrs)->addr_bytes[0]), \
+                     ((mac_addrs)->addr_bytes[1]), \
+                     ((mac_addrs)->addr_bytes[2]), \
+                     ((mac_addrs)->addr_bytes[3]), \
+                     ((mac_addrs)->addr_bytes[4]), \
+                     ((mac_addrs)->addr_bytes[5])
+
+static const struct rte_eth_conf default_port_conf = {
+    .rxmode = {
+        .max_rx_pkt_len = ETHER_MAX_LEN,
```
Comment (pengjianzhang) [inline src/virtio-user.c:55]: compile error at DPDK 21.11.1

src/virtio-user.c:55:10: error: ‘struct rte_eth_rxmode’ has no member named ‘max_rx_pkt_len’; did you mean ‘max_lro_pkt_size’?
   55 |         .max_rx_pkt_len = ETHER_MAX_LEN,
      |          ^~~~~~~~~~~~~~
      |          max_lro_pkt_size
src/virtio-user.c:56:46: error: ‘DEV_RX_OFFLOAD_JUMBO_FRAME’ undeclared here (not in a function); did you mean ‘DEV_RX_OFFLOAD_TIMESTAMP’?
   56 |         .offloads = DEV_RX_OFFLOAD_TCP_LRO | DEV_RX_OFFLOAD_JUMBO_FRAME,
      |                                              ^~~~~~~~~~~~~~~~~~~~~~~~~~
      |                                              DEV_RX_OFFLOAD_TIMESTA

File: src/virtio-user.c [hunk 4]
```diff
@@ -0,0 +1,372 @@
...
+
+
+static inline int
+configure_vdev(uint16_t port_id, struct rte_mempool *mb_pool)
+{
+	struct rte_ether_addr addr;
+	int ret;
+
+	if (!rte_eth_dev_is_valid_port(port_id))
+		return -1;
```
Comment (pengjianzhang) [inline src/virtio-user.c:91]: if/else needs to use {}

File: src/virtio-user.c [hunk 5]
```diff
@@ -0,0 +1,372 @@
...
+        snprintf(name, VDEV_NAME_SIZE, "%sc%1d", cfg->kni_ifname, idx);
+    }
+}
+
+
+static inline int
+configure_vdev(uint16_t port_id, struct rte_mempool *mb_pool)
+{
+	struct rte_ether_addr addr;
+	int ret;
```
Comment (pengjianzhang) [inline src/virtio-user.c:88]: Variables are sorted from small to large according to their sizeof(type)

File: src/virtio-user.c [hunk 6]
```diff
@@ -0,0 +1,372 @@
...
+
+	if (!rte_eth_dev_is_valid_port(port_id))
+		return -1;
+
+	ret = rte_eth_dev_configure(port_id, QUEUE_NUM, QUEUE_NUM,&default_port_conf);
+	if (ret != 0)
+		rte_exit(EXIT_FAILURE, "dev config failed\n");
```
Comment (pengjianzhang) [inline src/virtio-user.c:95]: {}

File: src/virtio-user.c [hunk 7]
```diff
@@ -0,0 +1,372 @@
...
+
+    int i = 0;
```
Comment (pengjianzhang) [inline src/virtio-user.c:97]: The variable is defined in the front part of the function

File: src/virtio-user.c [hunk 8]
```diff
@@ -0,0 +1,372 @@
...
+    for (; i < QUEUE_NUM; ++i) {
+        ret = rte_eth_tx_queue_setup(port_id, i, TX_DESC_PER_QUEUE,
+                rte_eth_dev_socket_id(port_id), NULL);
+        if (ret < 0)
+        {
```
Comment (pengjianzhang) [inline src/virtio-user.c:102]: if ( xxx ) {

}

File: src/virtio-user.c [hunk 9]
```diff
@@ -0,0 +1,372 @@
...
+            rte_exit(EXIT_FAILURE, "queue setup failed\n");
+        }
+            
+        ret = rte_eth_rx_queue_setup(port_id, i, RX_DESC_PER_QUEUE,
+                rte_eth_dev_socket_id(port_id), NULL, mb_pool);
+        if (ret < 0)
+        {
```
Comment (pengjianzhang) [inline src/virtio-user.c:109]: if (xxx) {
}

File: src/virtio-user.c [hunk 10]
```diff
@@ -0,0 +1,372 @@
...
+            rte_exit(EXIT_FAILURE, "queue setup failed\n");
+        }
+    }
+        
+
+	ret = rte_eth_dev_start(port_id);
+	if (ret < 0)
```
Comment (pengjianzhang) [inline src/virtio-user.c:116]: if (xxx) {
}

File: src/virtio-user.c [hunk 11]
```diff
@@ -0,0 +1,372 @@
...
+    {
+        rte_exit(EXIT_FAILURE, "dev start failed\n");
+    }
+		
+
+	ret = rte_eth_macaddr_get(port_id, &addr);
+	if (ret != 0) {
+        rte_exit(EXIT_FAILURE, "macaddr get failed\n");
```
Comment (pengjianzhang) [inline src/virtio-user.c:124]: Alignment

Comment (pengjianzhang) [inline src/virtio-user.c:124]: Alignment

File: src/virtio-user.c [hunk 12]
```diff
@@ -0,0 +1,372 @@
...
+    }
+
+	printf("Port %u MAC: %02"PRIx8" %02"PRIx8" %02"PRIx8
```
Comment (pengjianzhang) [inline src/virtio-user.c:127]: don't printf

File: src/virtio-user.c [hunk 13]
```diff
@@ -0,0 +1,372 @@
...
+
+	return 0;
+}
+
+static void *kni_alloc(struct config *cfg, struct netif_port *port)
+{
+    void *kni = NULL;
+    uint16_t port_id;
+    char kernel_name[VDEV_NAME_SIZE];
+	char vdev_args[VDEV_NAME_SIZE];
```
Comment (pengjianzhang) [inline src/virtio-user.c:142]: sort by size
alignment

File: src/virtio-user.c [hunk 14]
```diff
@@ -0,0 +1,372 @@
...
+    }
+}
+
+void kni_recv(struct work_space *ws, struct rte_mbuf *m)
+{
+    struct netif_port *port = NULL;
+    void *kni = NULL;
+    struct rte_mbuf *mbufs[NB_RXD];
+    struct rte_ring *kr = NULL;
+    int i, cnt, n = 0, send_n=0;
```
Comment (pengjianzhang) [inline src/virtio-user.c:294]: Each variable occupies one line and is assigned an initial value.

File: src/virtio-user.c [hunk 15]
```diff
@@ -0,0 +1,372 @@
...
+
+    port = ws->port;
+    kni = port->kni;
+    kr = port->kni_ring;
+    /**
+     * core that holds queue 0 is in charge of this port's kni work
```
Comment (pengjianzhang) [inline src/virtio-user.c:300]: bad comment format 
```
/* 
 *
 * */
```

**Review by pengjianzhang — CHANGES_REQUESTED** [review]

**Comment by pengjianzhang** [comment]
The spaces at the end of the line need to be removed

**Comment by fkcrazy001** [comment]
才注意到评论。今天验证下这些问题

File: src/virtio-user.c (hunk 3 above)
Comment (fkcrazy001) [reply src/virtio-user.c]: 已修改，卸载特性实际不需要，已经删除。

File: src/virtio-user.c (hunk 4 above)
Comment (fkcrazy001) [reply src/virtio-user.c:84]: 已修改

File: src/virtio-user.c (hunk 5 above)
Comment (fkcrazy001) [reply src/virtio-user.c]: 已修改

**Comment by fkcrazy001** [comment]
在我的vscode中，似乎没有对齐问题，我再研究下

**Comment by pengjianzhang** [comment]
行尾还是有空格

File: src/virtio-user.c (hunk 4 above)
Comment (fkcrazy001) [reply src/virtio-user.c]: fixed

File: src/virtio-user.c (hunk 11 above)
Comment (fkcrazy001) [reply src/virtio-user.c:110]: > Alignment

fixed

Comment (fkcrazy001) [reply src/virtio-user.c:110]: > Alignment

fixed

File: src/virtio-user.c (hunk 15 above)
Comment (fkcrazy001) [reply src/virtio-user.c:278]: > bad comment format
> 
> ```
> /* 
>  *
>  * */
> ```

fixed

**Review by fkcrazy001 — COMMENTED** [review]
alignment issues fixed

**Comment by pengjianzhang** [comment]
一些格式问题，请查邮箱

**Comment by pengjianzhang** [comment]
还是有些行尾空格

**Comment by fkcrazy001** [comment]
已修改，将新的commit合并并重新提了一个pr。

https://github.com/baidu/dperf/pull/481

**Pull request closed without merging**